keps/sig-network/3705-cloud-node-ips/README.md

```diff
@@ -0,0 +1,638 @@
+cloud-provider-specific changes should be needed; we should be able to
+make the needed changes in the `cloud-provider` module, and then the
+individual providers just need to revendor to the new version.
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+##### Unit tests
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, for Alpha try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `<package>`: `<date>` - `<test coverage>`
+
+##### Integration tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- <test>: <link to test coverage>
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+```
+<<[UNRESOLVED e2e ]>>
+
+I'm not sure how we currently handle cloud-provider e2e. GCP does not
+support IPv6 and `kind` does not use a cloud provider, so we cannot
+test the new code/behavior in any of the "core" e2e tests.
```

Comment:
@elmiko You've been researching our Cloud Provider E2E tests, perhaps you can help here?

Reply:
there is a whole series of dual stack tests in the upstream k/k e2e's, they are usually annotated with `[sig-network] [Feature:IPv6DualStack]` in their description/hierarchy text. i would expect that once we have the dual stack support in place, i would think we could run those e2e's and they should confirm the behavior.

mind you, these tests would not specifically test the CCM, but would ensure that the dual stack behavior is operating as expected with respect to Service objects.  in the research i've been doing, these tests might require some changes to ensure we could gate them properly for infrastructures where ipv6 support exists. but, it [appears that we are already running these tests on some platforms](https://sippy.dptools.openshift.org/sippy-ng/tests/4.13?filters=%257B%2522items%2522%253A%255B%257B%2522columnField%2522%253A%2522current_runs%2522%252C%2522operatorValue%2522%253A%2522%253E%253D%2522%252C%2522value%2522%253A%25227%2522%257D%252C%257B%2522columnField%2522%253A%2522variants%2522%252C%2522not%2522%253Atrue%252C%2522operatorValue%2522%253A%2522contains%2522%252C%2522value%2522%253A%2522never-stable%2522%257D%252C%257B%2522id%2522%253A99%252C%2522columnField%2522%253A%2522name%2522%252C%2522operatorValue%2522%253A%2522contains%2522%252C%2522value%2522%253A%2522%255Bsig-network%255D%2520%255BFeature%253AIPv6DualStack%255D%2522%257D%255D%252C%2522linkOperator%2522%253A%2522and%2522%257D&sort=asc&sortField=current_working_percentage)

Reply:
it's also worth mentioning that we might need to write some LoadBalancer specific service tests to ensure the CCM functionality is actually working, but i think we would first need to figure out where we will run those tests will run since it sounds like GCP might not be sufficient for the ipv6 stuff.

Reply:
Yeah, I know there are dual-stack e2e _tests_; my question was about dual-stack e2e _jobs_ (that are not kind-on-GCP-based). We can't test the behavior of dual-stack `--node-ip` in clouds unless we have dual-stack e2e jobs in those clouds.

Reply:
(ah, I said it wrong in the quoted text above; I should have said 'we cannot
test the new code/behavior in any of the "core" e2e jobs.' not 'e2e tests')

Reply:
ack, apologies for the confusion @danwinship , i am currently working with sig cloud provider to expand the ccm specific test suite. what i would like to see is an expansion, or new, interface for cloud providers to assist in making these tests more generic from the test repo perspective, allowing us to use a cloud provider interface to gate action on those tests. i am still in the process of doing the research, and it will most likely result in a KEP. then the next step with regard to the dual-stack work would be enabling those tests to run on a platform that has support for ipv6 and has implemented the new interface.

that's what i would like to see happen, but we are still in the research phase for that work and it will most likely need a KEP and some code changes before we can have that as a possibility for these tests. i'm not sure what a short term workaround would be, but happy to collaborate.

Reply:
it should be possible to run a kind cluster with an external cloud provider 🤔

Reply:
I can help with that

Reply:
> it should be possible to run a kind cluster with an external cloud provider thinking

++ @aojea  it absolutely should be possible

Reply:
>  I can help with that

i'd love to collaborate if possible, i have a lot to learn about our testing infrastructure.